### PR TITLE
Implement fully mixed mode for output_stream-s

### PR DESCRIPTION
**Pull request by xemul**
Nowadays output_stream supports two write() options -- buffered and zero-copy. The former copies provided data into an internal buffer and the latter just grabs the buffers and stores them in internal vector. 

Combining both methods together is limitedly possible -- one may do zero-copy writes after buffered writes, but must flush and (!) wait for the flush to actually happen to make another buffered write again. The "waiting" can be complicated in case batch-flushing option is ON, so this mode is very delicate and must be used with care.

This PR makes the output_stream support intermixed buffered and zero-copy writes in any order without restrictions :crossed_fingers: . The PR also includes an exhaustive unit test that validates all combinations of buffered vs zero-copy write-s for 1 through 4 buffers of various sizes (below, equal and up to 3x times above the stream buffer size) for trim-to-size option on and off.

All patches are AI-assisted (Claude Sonnet 4.6)

### CODE REVIEW

**Comment by avikivity** [comment]
Not reviewed yet, but wow

File: include/seastar/core/iostream-impl.hh [hunk 1]
```diff
@@ -111,12 +111,16 @@ future<> output_stream<CharType>::write(std::span<temporary_buffer<CharType>> bu
     size_t size = std::accumulate(bufs.begin(), bufs.end(), size_t(0), [] (size_t s, const auto& b) { return s + b.size(); });
     if (size != 0) {
```
Comment (avikivity) [inline include/seastar/core/iostream-impl.hh:112]: > allocation), share the filled prefix as _buf.share(0, _end) and
  advance _buf past it with trim_front(_end). The same underlying

This is good independently of the changes, but it does cause a new allocation, for the reference count.

I think we should enhance the allocator to be able to return a (ptr, refcount_ptr) pair, with the refcount stored adjacent to the memory (for small allocations) or in struct page (for large allocations). IIRC the freebsd kernel has a similar facility.

Not for now.

File: include/seastar/core/iostream-impl.hh [hunk 2]
```diff
@@ -370,29 +383,69 @@ template <typename CharType>
...
     try {
-        SEASTAR_ASSERT(_zc_bufs.empty() && "Mixing buffered writes and zero-copy writes not supported yet");
         if (!_end && (n >= _size)) {
             temporary_buffer<char> tmp = _fd.allocate_buffer(n);
             std::copy(buf, buf + n, tmp.get_write());
+            if (!_zc_bufs.empty()) {
+            // No buffered data yet, but zero-copy data is pending.
+            // Append to _zc_bufs so ordering is preserved.
+            _zc_len += n;
+            _zc_bufs.emplace_back(std::move(tmp));
```
Comment (avikivity) [inline include/seastar/core/iostream-impl.hh]: No reason for broken indents here.

Comment (avikivity) [inline include/seastar/core/iostream-impl.hh]: I'm not sure it matters, but better to switch the order of these two lines for exception safety. One isn't supposed to call flush() after an exception, so the inconsistency likely doesn't matter, but better to be careful.

File: include/seastar/core/iostream-impl.hh [hunk 3]
```diff
@@ -370,29 +383,69 @@ template <typename CharType>
...
+            if (_zc_len >= _size) {
```
Comment (avikivity) [inline include/seastar/core/iostream-impl.hh]: Please document the invariants near _buf and _zc_bufs.

If I understand correctly, the logical state of the stream is now _zc_bufs followed by span(_buf.data(), _end). If a copying write comes next it is appended to _bug, if a zero copy write comes the used prefix of _buf is appended to _zc_bufs, then the new writes, then the reminder of _buf.

It may make sense to rename _zc_bufs to _prefix_bufs.

File: include/seastar/core/iostream-impl.hh [hunk 4]
```diff
@@ -419,12 +472,21 @@ void add_to_flush_poller(output_stream<char>& x) noexcept;
...
+        if (_zc_bufs.empty()) {
         _buf.trim(_end);
         _end = 0;
         return _fd.put(std::move(_buf)).then([this] {
             return _fd.flush();
         });
-    } else if (!_zc_bufs.empty()) {
+        } else {
+            // Fold buffered tail into the zero-copy vector and flush together.
+            _zc_bufs.emplace_back(_buf.share(0, _end));
```
Comment (avikivity) [inline include/seastar/core/iostream-impl.hh:482]: I wonder if using a small_vector\<temporary_buffer> where we append to the last vector would simplify things.

**Comment by avikivity** [comment]
Very good.

File: include/seastar/core/iostream-impl.hh (hunk 2 above)
Comment (xemul) [reply include/seastar/core/iostream-impl.hh]: Presumably, you're reviewing only [this patch](https://github.com/scylladb/seastar/pull/3285/changes/9e3b578a9af3f69a9d243586023bd552a8044a8d). The indentation is left broken deliberately to simplify the review (well -- my review from what AI generated), [next patch](https://github.com/scylladb/seastar/pull/3285/changes/3e57be4ec3f6a0f107c4ada7d383459cd5cba3a4) restores it

File: include/seastar/core/iostream-impl.hh (hunk 3 above)
Comment (xemul) [reply include/seastar/core/iostream-impl.hh]: Yes, that's how it works now. I'll add the code comment

File: include/seastar/core/iostream-impl.hh (hunk 2 above)
Comment (avikivity) [reply include/seastar/core/iostream-impl.hh]: Yes. But here you introduce an if and some code it controls, and the new code it controls is not indented.

No need to change.

File: include/seastar/core/iostream-impl.hh (hunk 3 above)
Comment (avikivity) [reply include/seastar/core/iostream-impl.hh]: It think it also makes sense to rename _size, it's too generic. Maybe call it _buffer_size.

**Comment by xemul** [comment]
upd:
- renamed _size -> _buffer_size
- added comment about new _buf/_zc_bufs lifecycle

File: include/seastar/core/iostream-impl.hh (hunk 3 above)
Comment (xemul) [reply include/seastar/core/iostream-impl.hh]: Fixed both in next version

File: include/seastar/core/iostream-impl.hh (hunk 4 above)
Comment (xemul) [reply include/seastar/core/iostream-impl.hh:482]: Not really. When switching from buffered write to zero-copy write there's a need to maintain two buffers -- the currently accumulated one (_buf) and its remnant if it wasn't filled. Currently this is done with the help of _buf itself and switching everything onto unified small_vector<temporary_buffer, 1> complicates this state management

Comment (avikivity) [reply include/seastar/core/iostream-impl.hh:482]: I think you're right.

File: include/seastar/core/iostream-impl.hh (hunk 2 above)
Comment (avikivity) [reply include/seastar/core/iostream-impl.hh]: This?

**Comment by xemul** [comment]
upd:
- made slow-write zero-copy path thread safer

File: include/seastar/core/iostream-impl.hh (hunk 2 above)
Comment (xemul) [reply include/seastar/core/iostream-impl.hh]: Fixed in next version. I believe there are more (preexisting) places like that, but I didn't check them carefully
